src/hotspot/os/linux/os_linux.cpp

```diff
@@ -6198,6 +6380,8 @@ void VM_Crac::doit() {
   // dry-run fails checkpoint
   bool ok = !_dry_run;
 
+  remap_old_imagedir();
```

Comment:
VM was not bothered the way CREngine saved the memory content. The mmaping is an implementation detail of the CR mechnism.

Have you considered switching off the mmaping in CRIU in this repeated checkpoint-restore sequence? Assuming we would be able communicate that to CREngine (in CRIU mmaping is an option).

Semantically, this patch propopses to handle a mapping twice, once in CRIU with mmaping and another time in the VM. There are some benefits of doing everything in the VM and having better control over the process. So it would be cleaner to do a practically big part of the memory management in the VM and leaving bootstraping only to the CRIU.

Reply:
I think that mmaping in CRIU is an important optimization that speeds up boot, so I did not want to force disabling that if you ever want to do the checkpoint again.

You're right that this mixes the abstractions and responsibilities. There's an alternative solution that would not require changes in the VM, but it's technically more complex: we could ptrace VM and replace the mapping for it externally (though I am not sure how exactly should we invoke syscall on behalf of the tracee - maybe a parasite code would be needed?). Since ptracing process needs elevated priviledges we should probably add this as a separate criu command that would be invoked by criuengine.
The advantage is clearer semantics and not relying on SIGSEGV handling here which some consider a sketchy practice. But it's a more complex solution.